### PR TITLE
unix: optimize uv_fs_readlink() memory allocation

### PR DESCRIPTION
**Pull request by bnoordhuis**
Libuv was allocating PATH_MAX+1 bytes to reserve space for the trailing
nul byte.

On platforms like Linux, where PATH_MAX is fixed at 4096, that meant we
were allocating two pages (as the page size is normally also 4096), even
though the second page was almost never used.

Change that to allocate PATH_MAX bytes and only resize when readlink()
actually writes that many bytes, which is practically never.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1045/

### CODE REVIEW

File: src/unix/fs.c [hunk 1]
```diff
@@ -388,29 +388,42 @@ static ssize_t uv__fs_pathmax_size(const char* path) {
...
     return -1;
   }
 
+  /* Uncommon case: resize to make room for the trailing nul byte. */
+  if (len == maxlen) {
+    newbuf = uv__realloc(buf, len + 1);
+
+    if (newbuf == NULL) {
+      uv__free(buf);
+      return -1;
```
Comment (santigimeno) [inline src/unix/fs.c]: Shouldn't we set `errno` to `ENOMEM` here?

Comment (bnoordhuis) [reply src/unix/fs.c]: Shouldn't be necessary. `realloc()` sets it and `uv_free()` is carefully written to leave errno untouched. We don't set it in many other places either.

Comment (santigimeno) [reply src/unix/fs.c]: > Shouldn't be necessary. `realloc()` sets it

Got it. Is this also the case for `malloc()`? Looking at the manual I'm not sure and we usually set  `errno` if `uv__malloc()` returns `NULL`.

Comment (bnoordhuis) [reply src/unix/fs.c]: `malloc()` always sets errno on failure (barring bugs in the allocator, of course.)

I can't find many examples where we set errno outside of src/unix/fs.c, we usually `return UV_ENOMEM` or `return UV__ERR(errno)`.

Comment (santigimeno) [reply src/unix/fs.c]: You're right, it wasn't that usual at all: I think I only account for 3. Thanks!

**Pull request closed without merging**